### PR TITLE
expander: add Expected/Actual newtypes

### PR DESCRIPTION
**Pull request by doyougnu**
fixes: #151

Hi @gelisam !

We don't know each other but I was talking to @david-christiansen and he mentioned this project. It seems extremely interesting and I would like to contribute. So I thought this would be a dead simple PR and issue to close out. Please let me know if there is any edits you'd like, I am more than happy to do them. For example, I can imagine `unExpected expected` could be confusing, so we could use `getExpected` for the projection instead.

### CODE REVIEW

File: src/Expander/Error.hs
Comment (gelisam) [inline src/Expander/Error.hs:72]: What's the purpose of this wrapper function? It looks like a smart constructor, but it doesn't check any condition, so I don't see why callers should use `expected` instead of `Expected`.

File: src/Expander.hs
Comment (doyougnu) [reply src/Expander.hs:1169]: absolutely! That sounds good to me.

**Comment by gelisam** [comment]
Nice, thanks for contributing! David and I happen to have a meeting about Klister in a few hours, I'll send you an invite, please join if you have time and we'll show you around the code!

File: src/Expander/Error.hs
Comment (doyougnu) [reply src/Expander/Error.hs:72]: Ah yes I forgot to mention this. The purpose of the helper function is to separate the constructor from the exported API. So should `Expected` or `Actual` ever change then all their call sites won't need to change because from the beginning of their existence we define and treat `expected` and `actual` as _the only way_ to create an `Actual` or `Expected`.

I felt that this was more in line with david's original suggestion because if we took this to a logical next step and started to `parse, don't validate` then these helpers would be the site to do the parsing since they are the functions that add the type witnesses to the type level.

We can totally remove them if you like. If it were just me I would hide the `Actual` and `Expected` constructors and only export the `actual` and `expected` functions to enforce the aforementioned API.

**Comment by doyougnu** [comment]
> Nice, thanks for contributing! David and I happen to have a meeting about Klister in a few hours, I'll send you an invite, please join if you have time and we'll show you around the code!
> 
> _edit_: the meeting has been moved to next week

Sounds great to me as I intend to do many more contributions. Is there a better avenue to chat about Klister than GH PRs and issue lists?

File: src/Expander/Error.hs
Comment (gelisam) [reply src/Expander/Error.hs:72]: Since Klister is an executable, not a library, I don't think it's worth jumping through hoops to keep the code forward- or backward-compatible, as there aren't that many call sites to update. But now that you've jumped through the hoop, I don't want to make you jump backwards through the same hoop either! It's fine either way. If you do want to keep the function wrappers though, let's go all the way and not export the data constructors, like you suggest.

I am a fan of the parse-don't-validate approach when it makes sense, but I don't see how that approach applies here, because there are no values which would fail to parse. The purpose of the newtype is simply to tag the two arguments so that the code looks wrong if the two arguments are swapped.

By the way, `requireDeclarationCat` is already using the parse-don't-validate approach, in that it parses a `MacroDest` into a `(DeclTreePtr, DeclOutputScopesPtr)`.

**Comment by gelisam** [comment]
> Is there a better avenue to chat about Klister than GH PRs and issue lists?

The team is very small, so we don't yet have a mailing list; feel free to email David and I.

The way we usually collaborate is indeed via issues and PRs (note the existence of the generic "discussion" label for issues which don't fit into the other categories; we use issues for a lot more than just bug reports), followed by a discussion on our monthly meeting (I sent you an invite), followed by the creation of a few extra issues to reflect the discussion's decisions.

If you are eager to jump in and can't wait for our next meeting, you and I could also schedule a video call via email.

File: src/Expander/TC.hs
Comment (gelisam) [inline src/Expander/TC.hs:264]: Huh, what's this `hey` doing there? Looks like a stray debugging line from an earlier PR, let's remove it while we're here

Comment (doyougnu) [reply src/Expander/TC.hs:264]: So responsive!

Comment (gelisam) [reply src/Expander/TC.hs]: Also, this `mismatch` function should probably ask for `Actual` and `Expected` arguments as well

File: src/Expander.hs
Comment (gelisam) [reply src/Expander.hs:1169]: ❌ This is the only blocker

**Review by gelisam — CHANGES_REQUESTED** [review]
Thanks again! I made a few comments here and there, but the only change which I think really matters is the fact that the calls to `expected` and `actual` make it clear that the existing code is incorrect, the expected and actual arguments are mistakenly swapped.

File: src/Expander.hs
Comment (doyougnu) [reply src/Expander.hs]: issued a fix in 028928fa490222deb657465a877d9dc2b9cc6bd4 but I want to write the small Klister program before merging.

**Review by gelisam — CHANGES_REQUESTED** [review]
It wasn't just `requireDeclarationCat` which was swapped, it was every single call site. That is, the constant value is the expected one, while the variable we get from the pattern match, typically called `other`, is the actual value we receive.

**Comment by doyougnu** [comment]
> It wasn't just `requireDeclarationCat` which was swapped, it was every single call sites. That is, the constant value is the expected one, while the variable we get from the pattern match, typically called `other`, is the actual value we receive.

aha! Of course. This is now a tomorrow task. Talk to you then.

File: src/Expander.hs
Comment (gelisam) [reply src/Expander.hs]: Here's a small Klister program which triggers the problem: `define` only works in a declaration context (such as the top-level) but here it is called inside `example`, which expects an expression.

```
#lang "prelude.kl"

(example (define x 5))
```

When I ran this program on the `main` branch, I was surprised to see that the error message is actually correct, meaning that the arguments are _not_ swapped after all!

```
$ cabal run klister -- run toy.kl 
#[toy.kl:3.10-3.22]<(define x 5)>:
  Used in a position expecting an expression
  but is valid in a position expecting a top-level declaration or example
ExitFailure 1
```

I think what is going on is that the error message is written from the perspective of the macro, rather than from the perspective of the syntactic category. That is, the same information could have been conveyed by the following error message, in which it is the expression syntactic category which is the actual value and the declaration syntactic category which is the expected value:

```
expected: an expression
found: declaration #[toy.kl:3.10-3.22]<(define x 5)>
```

Thus, I think that the `Actual` and `Expected` newtype are actually making things worse for this particular function, as it makes the correct code look wrong and the wrong code look correct :upside_down_face: 

I still think the `Actual` and `Expected` newtypes themselves are useful, but for other functions such as `mismatch`.

For `WrongSyntacticCategory`, perhaps it would make more sense to use newtypes with names like `DestinationSyntacticCategory` and `MacroSyntacticCategory`?

Comment (doyougnu) [reply src/Expander.hs]: Alright let's take stock of the path we've traveled:
1) We agree that a `Actual` and `Expected` newtypes are useful and should be included in the PR.
2) But, in contrast to the prior suspicion `requireDeclarationCat` is not actually buggy because the error message of `toy.kl` is correct.
3) So the tasks are to (a) revert the commit that swaps `requireDeclarationCat` and (b) add more newtypes that give a proper denotation in the domain, suggestions are `DestinationSyntacticCategory` and `MacroSyntacticCategory`

But I'm still stuck on (2), here is why. If we look at the code that generates that error message on `main` we find:

```hs
...
  pp env (WrongSyntacticCategory stx actual expected) =
    hang 2 $ group $
    vsep [ pp env stx <> text ":"
...
```

which indicates to me that argument 2 and argument 3 to `WrongSyntacticCategory` are indeed interpreted as `Actual` and `Expected` respectively. So then if we cross reference with `requireDeclarationCat` on `main` we get:

```
requireDeclarationCat :: Syntax -> MacroDest -> Expand (DeclTreePtr, DeclOutputScopesPtr)
requireDeclarationCat _ (DeclTreeDest dest outScopesDest) = return (dest, outScopesDest)
requireDeclarationCat stx other =
  throwError $ WrongSyntacticCategory stx DeclarationCat (problemCategory other)
```

which does indeed set `DeclarationCat` to the actual value and `problemCategory other` to the expected, which _does_ still seem wrong to me _even_ if the error message was correct. Is it possible that `requireDeclarationCat` was not even called? Or I wonder if we could verify the call stack to be sure. What is surprising to me is that if I run `toy.kl` on my branch then I still get:
```
20:15:10 ❯ ./result/bin/klister run examples/failing-examples/toy.kl 
#[toy.kl:5.10-5.22]<(define x 5)>:
  Used in a position expecting a top-level declaration or example
  but is valid in a position expecting an expression
ExitFailure 1
```
That is, the error message is the same, which implies to me that `requireDeclarationCat` is not in the code execution path because we should expect to see a change after I swapped the arguments, but didn't.

But besides that hypothesized bug, I'm not in favor of `DestinationSyntacticCategory` and `MacroSyntacticCategory` because it implies `Destination` as an opposite to `Macro`. To make clear what I mean I'd rather have `Destination SyntacticCategory` and `Source SyntacticCategory` but I am unsure if there are other `Source`s than `Macro`.

Comment (gelisam) [reply src/Expander.hs]: > on my branch then I still get:
> 
>     20:15:10 ❯ ./result/bin/klister run examples/failing-examples/toy.kl 
>     #[toy.kl:5.10-5.22]<(define x 5)>:
>       Used in a position expecting a top-level declaration or example
>       but is valid in a position expecting an expression
>     ExitFailure 1
> 
> That is, the error message is the same

Look closer! :)

The error message above is _not_ the same error message as the one which the `main` branch gives:

    $ cabal run klister -- run toy.kl 
    #[toy.kl:3.10-3.22]<(define x 5)>:
      Used in a position expecting an expression
      but is valid in a position expecting a top-level declaration or example
    ExitFailure 1

It is the error message on `main` which is correct, because `define` is a macro which is valid in a position expecting a top-level declaration or an example (that is, in the Declaration syntactic category), but it is used inside `(example ...)`, which expects an expression.

You are right that `pp` uses the order (actual, expected), while `requireDeclarationCat` uses the order (expected, actual). Surprisingly, both are correct, they just have a different interpretation of what "actual" and "expected" mean! `pp` is taking the point of view of the `(example ...)` context: that context _expects_ the `...` to be an expression, but the _actual_ `...` is `define`, a definition. `requireDeclarationCat` takes the point of view of the `define` macro: that macro _expects_ to be called in a definition context, but the _actual_ context is `(example ...)`, which is an expression context. That's why I would prefer something like `Context` and `Macro` instead of `Actual` and `Expected`.

> But besides that hypothesized bug, I'm not in favor of `DestinationSyntacticCategory` and `MacroSyntacticCategory` because it implies `Destination` as an opposite to `Macro`. To make clear what I mean I'd rather have `Destination SyntacticCategory` and `Source SyntacticCategory` but I am unsure if there are other `Source`s than `Macro`.

I'm not too happy with the names `DestinationSyntacticCategory` and `MacroSyntacticCategory` either, thanks for suggesting something better! I like how `Source` and `Destination` are very general, like `Actual` and `Expected`. I'm not sure these are quite the right terms here though, because we aren't talking about a path which goes from a source to a destination. In this case, the word "destination" describe a hole, or a location, which gets filled by a value. Maybe `Hole` and `Plug`? `Caller` and `Callee`? `Parameter` and `Argument`?

Comment (gelisam) [reply src/Expander.hs:1169]: Maybe `Socket` and `Plug`?

Comment (doyougnu) [reply src/Expander.hs:1169]: Yes! That sounds perfect. I liked hole and fill because it fits the construction or carpentry motif (stuck, adhesive and bind)

Comment (gelisam) [reply src/Expander.hs:1169]: [Mortise and Tenon](https://en.m.wikipedia.org/wiki/Mortise_and_tenon), then?

File: src/Expander.hs
Comment (gelisam) [inline src/Expander.hs:1169]: This is again backwards... `requireDeclarationCat` checks if the _destination_ expects a declaration, so `other` is the `mortise`, not the `tenon`.

Comment (doyougnu) [reply src/Expander.hs:1169]: :face_exhaling: good thing the type system will eventually encode this because I keep messing it up!

File: src/Expander/Error.hs
Comment (gelisam) [inline src/Expander/Error.hs:206]: Also backwards: this part of the error message is talking about what the position expects, so that's the mortise, while the other part is talking about what we're putting in that position, so that's the tenon.

**Comment by doyougnu** [comment]
> All good now!

let me manually squash to remove the noise in the commit history

**Comment by doyougnu** [comment]
Ack I've become too used to gitlab. I've set my primary email in github so that my name doesn't show up as both `doyougnu` and `Jeffrey Young` in the commit log.

**Pull request merged**
